AmplifyPlugins/API/AWSAPICategoryPlugin/APIError+Extension.swift

```diff
@@ -0,0 +1,33 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+import AppSyncRealTimeClient
+
+extension APIError {
+
+    static let UnauthorizedMessageString: String = "Unauthorized"
+
+    /// Consolidates unauthorized error scenarios coming from `APIError.httpStatusError` and `APIError.operationError`.
+    /// For `.httpStatusError`, this checks if the status code is 403. For `.operationError`, this checks if the error
+    /// description contains "Unauthorized".
+    ///
+    /// **Warning** Customized server responses that indicate unauthorized may not match the internal mapping done
+    /// in this API and return `false`. Check APIError enum cases directly.
+    ///
+    /// - Returns: `true` if unauthorized error, `false` otherwise
+    public func isUnauthorized() -> Bool {
+        if case .operationError(let errorDescription, _, _) = self,
+           errorDescription.range(of: APIError.UnauthorizedMessageString, options: .caseInsensitive) != nil {
+            return true
+        } else if case .httpStatusError(let statusCode, _) = self, statusCode == 403 {
+            return true
+        }
+
+        return false
+    }
```

Comment:
I see that you call out in the code documentation that this logic may not match customized responses. Can we try to allow custom rules without sacrificing DX?
e.g.
```suggestion
       private var is403: Bool {
        switch self {
        case .httpStatusError(let statusCode, _):
            return statusCode == 403
        default:
            return false
        }
    }
    
    private var containsUnauthorizedMessageString: Bool {
        switch self {
        case .operationError(let errorDescription, _, _):
            return errorDescription.range(of: "Unauthorized", options: .caseInsensitive) != nil
        default: return false
        }
    }
    
    public struct UnauthorizedDetermining {
        let isUnauthenticated: (APIError) -> Bool
        
        public init(isUnauthenticated: @escaping (APIError) -> Bool) {
            self.isUnauthenticated = isUnauthenticated
        }
        
        public static let `default` = UnauthorizedDetermining { error in
            error.containsUnauthorizedMessageString || error.is403
        }
    }
    
    public func isUnauthorized(rule: UnauthorizedDetermining = .default) -> Bool {
        rule.isUnauthenticated(self)
    }
```

Reply:
Awesome addition of `UnauthorizedDetermining`, thanks for the code snippet, I incorporated this into the latest commit